File: pages/files/files.js
// When the mouse enters an editable folder item, change its icon to the edit
// icon.
up.on('mouseover', '.folder[editable],.document[editable]', (evt, elm) => {
  const icon = up.element.get(elm, 's-icon')
  if (!icon.hasAttribute('icon-save')) icon.setAttribute('icon-save', icon.getAttribute('icon'))
  icon.setAttribute('icon', 'edit')
})
// When the mouse leaves an editable folder item, change its icon back to its
// original icon.
up.on('mouseout', '.folder[editable],.document[editable]', (evt, elm) => {
  const icon = up.element.get(elm, 's-icon')
  if (icon.hasAttribute('icon-save')) icon.setAttribute('icon', icon.getAttribute('icon-save'))
})
// When someone clicks on the icon of an editable folder, open the folder edit
// dialog in a layer.
up.on('click', '.folder[editable] svg', (evt, elm) => {
  const folder = elm.closest('.folder')
  up.layer.open({
    url: `/folderedit/${folder.dataset.id}`,
    history: false,
    layer: 'new',
    size: 'grow',
    dismissable: 'key',
  })
})
// When someone clicks on the icon of an editable document, open the document
// edit dialog in a layer.
up.on('click', '.document[editable] s-icon', (evt, elm) => {
  const doc = elm.closest('.document')
  const folder = doc.closest('.folderContents').previousElementSibling
  up.layer.open({
    url: `/docedit/${folder.dataset.id}/${doc.dataset.id}`,
    history: false,
    layer: 'new',
    size: 'grow',
    dismissable: 'key',
  })
})
// When someone selects a file in the document edit dialog, and the file name
// field is empty, put the name of the selected file into the file name field.
up.on('input', '#doceditFile', (evt, elm) => {
  let filename = elm.value
  let idx = filename.lastIndexOf('/')
  if (idx >= 0) filename = filename.substr(idx + 1)
  idx = filename.lastIndexOf('\\')
  if (idx >= 0) filename = filename.substr(idx + 1)
  if (!filename) return
  const namefield = document.getElementById('doceditName')
  if (!namefield.value) namefield.value = filename
})
// When a folder item is added with the "folderItem-new" class, which means it
// was just added to the folder, pause a moment and then remove that class.
// This causes the new item background color to fade away.
up.compiler('.folderItem-new', { batch: true }, elms => {
  setTimeout(() => {
    elms.forEach(elm => { elm.classList.remove('folderItem-new') })
  }, 0)
})
  ; (function () {
    // containingFolder returns the .folder element for the folder that contains
    // the supplied element (a .folder or .document).
    function containingFolder(elm) {
      elm = elm.closest('.folderContents')
      if (elm) elm = elm.previousElementSibling
      return elm
    }
    // canDrop returns whether the toparent folder can receive the item that is
    // being dragged (as identified in the supplied event).
    function canDrop(evt, toparent) {
      return evt.dataTransfer.types.some(t => {
        // If the item that is being dragged is a URL or one or more files, any
        // folder can receive those.
        if (t === 'Files' || t === 'text/uri-list') return true
        // If the item being dragged is a document, it can be received as long
        // as it's not being dragged to the folder it's already in.  The folder
        // it's currently in is encoded in the data transfer type.
        const docmatch = t.match(/^x-serv-document;folder=(\d+)$/)
        if (docmatch && docmatch[1] !== toparent.dataset.id) return true
        // If the item being dragged is a folder, it can be received as long as
        // it's not being dragged to the folder it's already in, to itself, or
        // to any descendant of itself.  Its ID and its parent ID are encoded in
        // the data transfer type.
        const foldermatch = t.match(/^x-serv-folder;folder=(\d+);parent=(\d+)$/)
        if (foldermatch) {
          let movingID = foldermatch[1]
          let fromparentID = foldermatch[2]
          // If the destination folder is the parent of the folder that's moving
          // — i.e., nothing's changing — deny the move.
          if (fromparentID === toparent.dataset.id) return false
          // If the destination folder is, or is descended from, the folder
          // that's moving, deny the move.
          let f = toparent
          while (f) {
            if (movingID === f.dataset.id) return false
            f = containingFolder(f)
          }
          return true
        }
        return false
      })
    }
    // canDelete returns whether the trash target can receive the item that is
    // being dragged (as identified in the supplied event).
    function canDelete(evt) {
      return evt.dataTransfer.types.some(t => t.startsWith('x-serv-document;') || t.startsWith('x-serv-folder;'))
    }
    // When we start dragging a folder, set the data transfer type, embedding
    // its ID and its parent's ID.  Also hide the Add buttons and, if the item
    // is deletable, show the trash drop target.
    up.on('dragstart', '.folder[draggable]', (evt, elm) => {
      evt.dataTransfer.setData(`x-serv-folder;folder=${elm.dataset.id};parent=${containingFolder(elm).dataset.id}`, elm.dataset.id)
      document.getElementById('folderButtons').style.display = 'none'
      if (elm.hasAttribute('deletable')) document.getElementById('folderDelete').style.display = null
    })
    // When we start dragging a document, set the data transfer type, embedding
    // its containing folder's ID.  Also hide the Add buttons and show the trash
    // drop target.
    up.on('dragstart', '.document[draggable]', (evt, elm) => {
      evt.dataTransfer.setData(`x-serv-document;folder=${containingFolder(elm).dataset.id}`, elm.dataset.id)
      document.getElementById('folderButtons').style.display = 'none'
      document.getElementById('folderDelete').style.display = null
    })
    // When we stop dragging something, hide the trash drop target and show the
    // Add buttons.
    up.on('dragend', '.folder[draggable], .document[draggable]', () => {
      document.getElementById('folderButtons').style.display = null
      document.getElementById('folderDelete').style.display = 'none'
    })
    // When something is being dragged over a folder that can receive it, tell
    // the browser that by preventing the dragover event.
    up.on('dragover', '.folder[editable]', (evt, elm) => {
      if (canDrop(evt, elm)) evt.preventDefault()
    })
    // When something deletable is being dragged over the trash drop target,
    // tell the browser that by preventing the dragover event.
    up.on('dragover', '#folderDelete', (evt, elm) => {
      if (canDelete(evt)) evt.preventDefault()
    })
    // When dragging into or out of an editable folder (or the icon or link
    // within the .folder element), add or remove the highlight on that folder.
    up.on('dragenter dragleave', '.folder[editable], .folder[editable] *', (evt, elm) => {
      folder = elm.closest('.folder')
      // Don't remove the highlight if the drag location is a different element
      // within the same .folder elemen.t
      if (evt.type === 'dragleave' && evt.relatedTarget && evt.relatedTarget.closest('.folder') === folder) return
      // Don't add the highlight if the target folder can't receive the item
      // being dragged.
      if (evt.type === 'dragenter' && !canDrop(evt, folder)) return
      // Toggle the highlight.
      folder.classList.toggle('folder-dragging', evt.type === 'dragenter')
    })
    // When dragging into or out of the trash drop target (or the icon within
    // it), add or remove the highlight on that target.
    up.on('dragenter dragleave', '#folderDelete, #folderDelete s-icon', (evt, elm) => {
      target = document.getElementById('folderDelete')
      // Don't add the highlight if the target folder can't receive the item
      // being dragged.
      if (evt.type === 'dragenter' && !canDelete(evt)) return
      // Toggle the highlight.
      target.classList.toggle('folder-dragging', evt.type === 'dragenter')
    })
    // Handle a drop of an item onto a folder.
    up.on('drop', '.folder[editable]', (evt, elm) => {
      if (!canDrop(evt, elm)) return
      let form
      evt.dataTransfer.types.some(t => {
        if (t.startsWith('x-serv-document')) {
          // Dropping a document.  Fill in that field of the drop form.
          const field = document.getElementById('folderDropDoc')
          field.value = evt.dataTransfer.getData(t)
          form = field.form
        }
        if (t.startsWith('x-serv-folder')) {
          // Dropping a folder.  Fill in that field of the drop form.
          const field = document.getElementById('folderDropFolder')
          field.value = evt.dataTransfer.getData(t)
          form = field.form
        }
        if (t === 'text/uri-list') {
          // Dropping a URL.  Fill in that field of the drop form.
          const url = evt.dataTransfer.getData('text/uri-list').replace(/\r\n.*/, '')
          const field = document.getElementById('folderDropURL')
          field.value = url
          form = field.form
        }
        if (t === 'Files') {
          // Dropping a set of files.  Fill in that field of the drop form.
          const field = document.getElementById('folderDropFiles')
          field.files = evt.dataTransfer.files
          form = field.form
        }
      })
      if (form) {
        // We have something to drop.  Set the form action based on the folder
        // being dropped, and submit the form.  The submission counts as a
        // browser history change only if it is a new URL.
        evt.preventDefault()
        form.action = elm.dataset.path
        up.submit(form, { history: form.action !== document.getElementById('folderpath').value })
      }
    })
    // Handle a drop of an item onto the trash target.
    up.on('drop', '#folderDelete', (evt, elm) => {
      if (!canDelete(evt)) return
      let form, history
      let action = document.getElementById('folderpath').value
      evt.dataTransfer.types.some(t => {
        if (t.startsWith('x-serv-document')) {
          // Dropping a document.  Fill in that field of the drop form.
          const field = document.getElementById('folderTrashDoc')
          field.value = evt.dataTransfer.getData(t)
          form = field.form
          history = false
        }
        if (t.startsWith('x-serv-folder')) {
          // Dropping a folder.  Fill in that field of the drop form.
          const field = document.getElementById('folderTrashFolder')
          field.value = evt.dataTransfer.getData(t)
          form = field.form
          if (evt.dataTransfer.getData(t) === document.getElementById('folderid').value) {
            action = action.replace(/\/[^/]*$/, '')
            history = true
          }
        }
      })
      if (form) {
        // We have something to drop.  Set the form action based on the item
        // being dropped, and submit the form.  The submission counts as a
        // browser history change only if it is a new URL.
        evt.preventDefault()
        form.action = action
        up.submit(form, { history })
      }
    })
  })()

File: pages/files/files.go
package files

import (
	"fmt"
	"io/fs"
	"net/http"
	"net/url"
	"os"
	"path"
	"path/filepath"
	"strings"

	"sunnyvaleserv.org/portal/pages/errpage"
	"sunnyvaleserv.org/portal/server/auth"
	"sunnyvaleserv.org/portal/store/document"
	"sunnyvaleserv.org/portal/store/folder"
	"sunnyvaleserv.org/portal/store/person"
	"sunnyvaleserv.org/portal/ui"
	"sunnyvaleserv.org/portal/util/htmlb"
	"sunnyvaleserv.org/portal/util/request"
)

// FolderFields is the set of Folder fields that need to be retrieved in Folder
// objects passed to the GetFolder function.
const FolderFields = folder.FID | folder.FName | folder.FURLName | folder.FViewer | folder.FEditor

// Handle handles /files/${path} requests.
func Handle(r *request.Request) {
	var (
		flist   []*folder.Folder
		docname string
		user    *person.Person
	)
	user = auth.SessionUser(r, 0, false)
	// Try to find the folder with that path.
	if flist, docname = folder.WithPath(r, r.Path, FolderFields); flist == nil {
		errpage.NotFound(r, user)
		return
	}
	// Make sure we have view privilege on the leaf folder.
	if !user.HasPrivLevel(flist[len(flist)-1].Viewer()) {
		if user == nil { // not logged in, so send a 401
			auth.SessionUser(r, 0, true)
		} else { // logged in, so send a 403
			errpage.Forbidden(r, user)
		}
		return
	}
	// Honor their document or folder request.
	if docname != "" {
		getDocument(r, user, flist[len(flist)-1], docname)
	} else if r.Method == http.MethodPost {
		postFolder(r, user, flist)
	} else {
		GetFolder(r, user, flist, 0, nil)
	}
}

// getDocument handles GET /files/${path}/${docname}.  Permissions have already
// been checked.
func getDocument(r *request.Request, user *person.Person, f *folder.Folder, docname string) {
	var (
		doc  *document.Document
		fh   *os.File
		stat fs.FileInfo
		err  error
	)
	if doc = document.WithName(r, f.ID(), docname); doc == nil || doc.URL != "" {
		errpage.NotFound(r, user)
		return
	}
	fh = document.Open(doc.ID)
	if stat, err = fh.Stat(); err != nil {
		panic(err)
	}
	switch strings.ToLower(filepath.Ext(docname)) {
	case ".jpeg", ".jpg", ".png", ".pdf":
		r.Header().Set("Content-Disposition", "inline")
	default:
		r.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=%q", docname))
	}
	r.Header().Set("Cache-Control", "no-cache")
	http.ServeContent(r, r.Request, doc.Name, stat.ModTime(), fh)
}

// postFolder handles POST /files/${path} when the path is a folder.  This is a
// post from a hidden form, used when something is dragged and dropped on the
// folder.
func postFolder(r *request.Request, user *person.Person, flist []*folder.Folder) {
	var (
		newfolder folder.ID
		newdocs   map[document.ID]bool
		f         = flist[len(flist)-1]
	)
	if !user.HasPrivLevel(f.Editor()) {
		errpage.Forbidden(r, user)
		return
	}
	if !auth.CheckCSRF(r, user) {
		return
	}
	if idstr := r.FormValue("doc"); idstr != "" {
		// They dropped a document from another folder onto this one.
		docid, ok := handleDropDocument(r, user, f, idstr)
		if !ok {
			return
		}
		if docid != 0 {
			newdocs = make(map[document.ID]bool)
			newdocs[docid] = true
		}
	}
	if idstr := r.FormValue("deldoc"); idstr != "" {
		// They dropped a document onto the trash target.
		if !handleDeleteDocument(r, user, idstr) {
			return
		}
	}
	if idstr := r.FormValue("folder"); idstr != "" {
		// They dropped another folder onto this one.
		folderid, ok := handleDropFolder(r, user, f, idstr)
		if !ok {
			return
		}
		newfolder = folderid
	}
	if idstr := r.FormValue("delfolder"); idstr != "" {
		// They dropped a folder onto the trash target.
		folderid, ok := handleDeleteFolder(r, user, idstr)
		if !ok {
			return
		}
		if folderid == f.ID() {
			flist = flist[:len(flist)-1]
		}
	}
	if u := r.FormValue("url"); u != "" {
		// They dropped a URL onto this folder.
		docid := handleDropURL(r, f, u)
		newdocs = map[document.ID]bool{docid: true}
	}
	if r.MultipartForm != nil && r.MultipartForm.File != nil && len(r.MultipartForm.File["file"]) != 0 {
		// They dropped one or more files onto this folder.
		files := r.MultipartForm.File["file"]
		newdocs = make(map[document.ID]bool)
		for _, file := range files {
			docid := handleDropFile(r, f, file)
			newdocs[docid] = true
		}
	}
	GetFolder(r, user, flist, newfolder, newdocs)
}

// GetFolder handles GET /files/${path} when the path is a folder.  Permissions
// have already been checked.
func GetFolder(r *request.Request, user *person.Person, flist []*folder.Folder, newfolder folder.ID, newdocs map[document.ID]bool) {
	var (
		location         string
		anythingEditable bool
	)
	// Notify the unpoly library in the client that this response is
	// GETtable, even if it is being sent as the response to a POST, i.e.,
	// after a drag/drop operation or an edit dialog submission.
	for _, f := range flist {
		if location == "" {
			location = "/files"
		} else {
			location = path.Join(location, f.URLName())
		}
	}
	r.Header().Set("X-Up-Method", "GET")
	r.Header().Set("X-Up-Location", location)
	// Now display the folder page.
	ui.Page(r, user, ui.PageOpts{Title: r.Loc("Files")}, func(main *htmlb.Element) {
		var fdiv = main
		var fpath string

		// Display each of the ancestor folders, up to and including the
		// target folder.
		for i, f := range flist {
			var label string

			if fpath == "" && i == len(flist)-1 && user == nil {
				fpath = "/"
				label = "Sunnyvale SERV"
			} else if fpath == "" {
				fpath = "/files"
				label = f.Name()
			} else {
				fpath = path.Join(fpath, f.URLName())
				label = f.Name()
			}
			canEdit := user.HasPrivLevel(f.Editor())
			canDelete := i == len(flist)-1 && canEdit && !folder.ExistsWithParent(r, f.ID()) && !document.ExistInFolder(r, f.ID())
			canDrag := canDelete || (i > 0 && user.HasPrivLevel(flist[i-1].Editor()))
			anythingEditable = anythingEditable || canEdit || canDelete
			fline := fdiv.E("div class=folder data-id=%d", f.ID(),
				canEdit, "editable data-path=%s", fpath,
				canDelete, "deletable",
				canDrag, "draggable=true")
			fline.E("s-icon icon=folder-open")
			if user.HasPrivLevel(f.Viewer()) {
				fline.E("a href=%s up-target=main", fpath, canDrag, "draggable=false").T(label)
			} else {
				fline.T(label)
			}
			fdiv = fdiv.E("div class=folderContents")
			if fpath == "/" {
				fpath = "/files"
			}
		}
		f := flist[len(flist)-1]
		canEdit := user.HasPrivLevel(f.Editor())
		fdiv.E("input type=hidden id=folderpath value=%s", fpath)
		fdiv.E("input type=hidden id=folderid value=%d", int(f.ID()))
		// Display the child folders of the target folder.
		folder.AllWithParent(r, f.ID(), FolderFields, func(cf *folder.Folder) {
			if user.HasPrivLevel(cf.Viewer()) {
				canEditChild := user.HasPrivLevel(cf.Editor())
				canDeleteChild := canEditChild && !folder.ExistsWithParent(r, cf.ID()) && !document.ExistInFolder(r, cf.ID())
				anythingEditable = anythingEditable || canEditChild || canDeleteChild
				fdiv.E("div class=folder data-id=%d", cf.ID(),
					cf.ID() == newfolder, "class=folderItem-new",
					canEditChild, "editable data-path=%s", path.Join(fpath, cf.URLName()),
					canDeleteChild, "deletable",
					canEdit, "draggable=true").
					E("s-icon icon=folder").P().
					E("a href=%s up-target=main", path.Join(fpath, cf.URLName()), canEdit, "draggable=false").
					T(cf.Name())
			}
		})
		// Display the documents in the target folder.
		document.AllInFolder(r, f.ID(), func(doc *document.Document) {
			var newtab bool
			ddiv := fdiv.E("div class=document data-id=%d", doc.ID,
				newdocs != nil && newdocs[doc.ID], "class=folderItem-new",
				canEdit, "editable deletable draggable=true")
			if doc.URL != "" {
				ddiv.E("s-icon icon=link")
				ddiv.E("a href=%s target=_blank", doc.URL, canEdit, "draggable=false").T(doc.Name)
				return
			}
			switch strings.ToLower(filepath.Ext(doc.Name)) {
			case ".pdf":
				newtab = true
				ddiv.E("s-icon icon=pdf")
			case ".png", ".jpeg", ".jpg":
				newtab = true
				ddiv.E("s-icon icon=image")
			case ".docx", ".doc":
				ddiv.E("s-icon icon=word")
			case ".pptx", ".ppt":
				ddiv.E("s-icon icon=powerpoint")
			case ".xlsx", ".xls":
				ddiv.E("s-icon icon=excel")
			default:
				ddiv.E("s-icon icon=file")
			}
			ddiv.E("a href=%s", path.Join(fpath, url.PathEscape(doc.Name)),
				newtab, "target=_blank",
				canEdit, "draggable=false").
				T(doc.Name)
		})
		// If the target folder is editable, display the add buttons.
		if canEdit {
			buttons := main.E("div id=folderButtons class=formButtons")
			buttons.E("a href=/docedit/%d/NEWFILE class='sbtn sbtn-primary' up-layer=new up-size=grow up-dismissable=key up-history=false>Add File", f.ID())
			buttons.E("a href=/docedit/%d/NEWURL class='sbtn sbtn-primary' up-layer=new up-size=grow up-dismissable=key up-history=false>Add Web Link", f.ID())
			buttons.E("a href=/folderedit/NEW?parent=%d class='sbtn sbtn-primary' up-layer=new up-size=grow up-dismissable=key up-history=false>Add Folder", f.ID())
			deltarget := main.E("div id=folderDelete style=display:none")
			deltarget.E("s-icon icon=trash")
			deltarget.R("Drag here to delete.")
		}
		// If any folder on the page is editable, add the hidden form
		// that's used by the client code to handle drag and drop.
		if anythingEditable {
			form := main.E("form method=POST enctype=multipart/form-data up-target=main style=display:none")
			form.E("input type=hidden name=csrf value=%s", r.CSRF)
			form.E("input type=file id=folderDropFiles name=file")
			form.E("input type=hidden id=folderDropURL name=url")
			form.E("input type=hidden id=folderDropDoc name=doc")
			form.E("input type=hidden id=folderDropFolder name=folder")
			form.E("input type=hidden id=folderTrashDoc name=deldoc")
			form.E("input type=hidden id=folderTrashFolder name=delfolder")
		}
	})
}
